File: is_valid_parentheses.py
def is_valid_parentheses():
    s = "[({})]"

    hm = {
        ")": "(",
        "]": "[",
        "}": "{"
    }

    if s and s[0] in hm:
        return False

    st = []
    for c in s:
        if c in hm:
            if st and st[-1] == hm[c]:
                st.pop()
            else:
                return False
        else:
            st.append(c)

    return len(st) == 0

print(is_valid_parentheses())

File: is_valid_parentheses.rs
use std::collections::HashMap;

fn is_valid_parentheses(s: String) -> bool {
    let mut hm = HashMap::new();
    hm.insert(')', '(');
    hm.insert(']', '[');
    hm.insert('}', '{');

    if s.starts_with(|c| hm.contains_key(&c)) {
        return false;
    }

    let mut st = Vec::new();

    for c in s.chars() {
        match hm.get(&c) {
            Some(&expected) => {
                if st.last() == Some(&expected) {
                    st.pop();
                } else {
                    return false;
                }
            }
            None => st.push(c),
        }
    }

    st.is_empty()
}

fn main() {
    let s = String::from("()[]{}");
    println!("Result: {:?}", is_valid_parentheses(s));
}
